File: apps/web/features/evaluation/schemas.ts
import { z } from "zod";

export const CaseEvaluationSchema = z.object({
  caseId: z.string(),
  query: z.string(),
  conditionKey: z.string(),
  expectedPatientId: z.string(),
  hit: z.boolean(),
  rank: z.number().int(),
  relevanceScore: z.number().nullable(),
  tookMs: z.number().int().nonnegative(),
  decoyLeaked: z.boolean(),
  topSnippet: z.string().nullable(),
});

export const EvaluationSummarySchema = z.object({
  totalCases: z.number().int().nonnegative(),
  hitCount: z.number().int().nonnegative(),
  hitRate: z.number(),
  mrr: z.number(),
  ndcg: z.number(),
  avgLatencyMs: z.number(),
  p50LatencyMs: z.number().int().nonnegative(),
  p95LatencyMs: z.number().int().nonnegative(),
  p99LatencyMs: z.number().int().nonnegative(),
  decoyLeakCount: z.number().int().nonnegative(),
});

export const EvaluationReportSchema = z.object({
  evaluatedAt: z.string(),
  summary: EvaluationSummarySchema,
  cases: z.array(CaseEvaluationSchema),
});

export type CaseEvaluation = z.infer<typeof CaseEvaluationSchema>;
export type EvaluationSummary = z.infer<typeof EvaluationSummarySchema>;
export type EvaluationReport = z.infer<typeof EvaluationReportSchema>;

File: services/api/app/features/evaluation/schemas.py
from app.schemas import CamelModel


class CaseEvaluation(CamelModel):
    case_id: str
    query: str
    condition_key: str
    expected_patient_id: str
    hit: bool
    rank: int  # 1-based; -1 if not found
    relevance_score: float | None = None
    took_ms: int
    decoy_leaked: bool
    top_snippet: str | None = None


class EvaluationSummary(CamelModel):
    total_cases: int
    hit_count: int
    hit_rate: float
    mrr: float  # Mean Reciprocal Rank
    avg_latency_ms: float
    decoy_leak_count: int


class EvaluationReport(CamelModel):
    evaluated_at: str
    summary: EvaluationSummary
    cases: list[CaseEvaluation]
